Headfluids base class (revision 2)

RonRahaman opened the pull request:
This introduces a `HeatFluidsDriver` base class from which `NekDriver` and `SurrogateHeatDriver` are derived.  

`HeatFluidsDriver` declares:
  * `double pressure_bc_`:  The inlet pressure.  A parameter `double pressure_bc` be passed to the constructor.
  * `density()`:  Returns a flattened xtensor of densities for each region.  The semantics of a "region" are left up to the derived class
  * `fluid_mask()`:  Returns a flattened xtensor of integers.  For each region, states whether that region is in fluid.  

These classes were also changed:  
* `NekDriver`:  
  * From `density()`, the fluid elements have meaningful densities and the solid elements have zero density.  Any coupled driver that uses the densities should also use `fluid_mask()` to filter the meaningful data.  
  * Has a new member, `pressure_` that is a scalar value read from the `<nek5000>` node in `enrico.xml`
* `OpenmcNekDriver`:
  * When instantiating the `NekDriver`, the system pressure from `enrico.xml` is passed to the `NekDriver` constructor.  
  * Has a new data member, `elem_densities_` that is updated from `NekDriver::density()`
  * Has a new data member `elem_fluid_mask_` that is instantiated from `NekDriver::fluid_mask()`
  * Has a new data member `cell_fluid_mask_` that is instantiated in `init_cell_mask()`.  
  * Has a new member function `update_cell_density()` that updates fluid densities based on `elem_densities_` and `cell_fluid_mask_`.   
  * Deleted the `pressure_` member
* `SurrogateHeatDriver`:  
  * `pressure_bc_` is not used but must still be passed to the constructor
  * `fluid_mask` is all zeros.  

TODO:
- [ ] Sanity check for simulation output
- [x] Update XML specs in docs

Revises #23 
Fixes #14 
Fixes #16

RonRahaman commented on include/enrico/nek_driver.h:
```diff
@@ -77,23 +76,12 @@ class NekDriver : public Driver {
-  //! \param global_elem  A global element ID
-  //! \return 1 if the global element is in fluid; 0 otherwise
-  int global_elem_is_in_fluid(int global_elem) const;
+  double temperature(int local_elem) const;
```
At one point, we discussed naming these functions:
``` cpp
double temperature_at(int local_elem) const;
```
To more-clearly distinguish them from
``` cpp
xt::xtensor<double, 1> temperature() const final;
```
What do you think?

paulromano replied:
I'm down for that. It would help to make it a bit clearer.

RonRahaman replied:
We could also define it as something like:
``` cpp
xt::tensor<doulbe, 1> temperature_at(const std:vector<int> &idx)
```

If so, we could include it in the `HeatFluidsDriver` base class.  

My reasoning is that `SurrogateHeatDriver` and `NekDriver` have different conventions for indexing regions, but this declaration could handle both.  You'd have an indexing scheme like `temperature_at({i,})` for `NekDriver` and `temperature_at({i, j, k})` for `SurrogateHeatDriver`

paulromano replied:
Doesn't seem like that's really needed as part of the heat-fluids driver interface though. `temperature()` is sufficient for that. I'm ok just having those be specific to the individual drivers at this point.

RonRahaman commented on src/openmc_nek_driver.cpp:
```diff
@@ -346,45 +391,56 @@ void OpenmcNekDriver::update_temperature()
 
         // Set temperature for cell instance
         average_temp /= total_vol;
+        Expects(average_temp > 0.0)
```
Not sure if you want `Expects` or `Ensures` here.

paulromano replied:
`Expects` is fine. You do need a semicolon at the end of this though.

RonRahaman replied:
`clang-tidy` actually complained that the semicolon is redundant, and I can safely omit it on both my machine and Travis.  But I looked at `gsl.h`, and it does seem like it's necessary in general:  

``` cpp
#if gsl_ELIDE_CONTRACT_EXPECTS
# define Expects( x )  /* Expects elided */
#elif gsl_CONFIG( CONTRACT_VIOLATION_THROWS_V )
# define Expects( x )  ::gsl::fail_fast_assert( (x), "GSL: Precondition failure at " __FILE__ ":" gsl_STRINGIFY(__LINE__) );
#else
# define Expects( x )  ::gsl::fail_fast_assert( (x) )
#endif
```
One of the macro definitions has a semicolon at the end, but the other doesn't.  They need to be consistent...

Anyway, I'll add the semicolon.

RonRahaman commented on src/openmc_nek_driver.cpp:
```diff
@@ -346,45 +391,56 @@ void OpenmcNekDriver::update_temperature()
-        if (any_in_fluid)
-          c.set_density(average_density / total_vol);
+        double density = average_density / total_vol;
+        Expects(density > 0.0)
```
Not sure if you want `Expects` or `Ensures` here.

paulromano replied:
Again, `Expects` is fine. Add a semicolon to be safe

paulromano commented on src/openmc_heat_driver.cpp:
```diff
@@ -18,7 +18,8 @@ OpenmcHeatDriver::OpenmcHeatDriver(MPI_Comm comm, pugi::xml_node node)
   openmc_driver_ = std::make_unique<OpenmcDriver>(comm);
   pugi::xml_node surr_node = node.child("heat_surrogate");
-  heat_driver_ = std::make_unique<SurrogateHeatDriver>(comm, surr_node);
+  double pressure = {node.child("pressure").text().as_double()};
```
Either `=` or `{ ... }`

paulromano commented on src/surrogate_heat_driver.cpp:
```diff
@@ -11,8 +11,10 @@
+SurrogateHeatDriver::SurrogateHeatDriver(MPI_Comm comm,
+                                         double pressure,
+                                         pugi::xml_node node)
+  : HeatFluidsDriver(comm, pressure)
```
Since `pressure_` is stored on `HeatFluidsDriver`, the check that `pressure > 0` should probably be contained there. Right now, it's only in `OpenmcNekDriver`.

RonRahaman replied:
Additionally,  should we remove the `pressure_` member from `OpenmcNekDriver`?  It is only used to instantiate a `NekDriver`.

paulromano replied:
Yes, good catch on that

paulromano commented:
It looks like moving the check on pressure broke the surrogate case (pressure is never specified for that case since it's not needed). I'd suggest just adding the pressure to the enrico.xml file for that case -- we'll probably want to use it later for the subchannel solver.

RonRahaman commented on include/enrico/heat_fluids_driver.h:
```diff
@@ -11,11 +11,23 @@ namespace enrico {
+  //! \return For each region, 1 if region is in fluid and 0 otherwise
+  virtual xt::xtensor<int, 1> fluid_mask() const = 0;
+
+  double pressure_;  //! System pressure in [MPa]
```
Would it make sense to declare `pressure_` as an `xarray<double>`?  In Nek, the pressure would be treated a 0D container, and another surrogate could treat it as a higher-dimensional container.  xtensor seems to have pretty consistent semantics for this.

aprilnovak replied:
If we move `pressure_` to `NekDriver` as in #31, then I think we can leave it as a scalar. 

If we add a pressure solution to `SurrogateHeatDriver`, then I don't think it's necessary to have the same pressure representation in `NekDriver` and `SurrogateHeatDriver` since `pressure_` would be private in `SurrogateHeatDriver`, similar to like `temperature_`. 

Nek's solution field is accessed by interface functions, while the `SurrogateHeatDriver`'s solution is made private so that we also are forced to access it through interfaces.

RonRahaman replied:
That makes sense.  I'll move `pressure_` back to `NekDriver`.  I'd like to do it in this PR instead of your separate one (#31), since they both touch a lot of the same stuff.

RonRahaman commented on include/enrico/nek_driver.h:
```diff
@@ -25,13 +25,17 @@ class NekDriver : public Driver {
   //!
   //! \param comm  The MPI communicator used to initialze Nek5000
-  explicit NekDriver(MPI_Comm comm, pugi::xml_node xml_root);
+  explicit NekDriver(MPI_Comm comm, double pressure, pugi::xml_node xml_root);
```
This makes sense.  We'll need to slightly change the specs of `enrico.xml`.  Currently, it looks like:

``` XML
<?xml version="1.0"?>
<stream>
  <driver_transport>openmc</driver_transport>
  <driver_heatfluids>nek5000</driver_heatfluids>
  <nek5000>
    <casename>rodcht</casename>
  </nek5000>
  <communication>overlapping</communication>
  <power>820.0</power>
  <pressure>12.7553</pressure>
  <max_timesteps>4</max_timesteps>
  <max_picard_iter>2</max_picard_iter>
  <openmc_procs_per_node>1</openmc_procs_per_node>
  <epsilon>0.5</epsilon>
</stream>
```

If we read pressure in `NekDriver`, we'll have to move `<pressure>` from the `<stream>` node to the `<nek5000>` node.  This is because the `<nek5000>` node is passed to the `NekDriver` constructor.

aprilnovak replied:
Oh good point. I think we should move it to the `nek5000` node, since it's the only solver that needs it.

RonRahaman commented on include/enrico/heat_fluids_driver.h:
```diff
@@ -11,11 +11,23 @@ namespace enrico {
 
-  //! Get the temperature in each element
-  //! \return Temperature in each element as [K]
+  virtual ~HeatFluidsDriver() = default;
```
Good to know!  We're only requiring C++14 in this project (see the `CMakeLists.txt`), so I don't really want to change this.  Also, putting virtual destructors in base classes was really drilled into me 15 years (!) ago, and I'll probably never stop doing it :P

RonRahaman commented:
After some conversations with @aprilnovak , we've moved the `pressure_` data member to `NekDriver`.  It is no longer in `OpenmcNekDriver` (as in the master branch); nor is it in `HeadFluidsDriver` (as it was earlier in this PR).  Furthermore, the `NekDriver` constructor does not take a `pressure` argument; rather, the pressure is initialized from the XML node that is passed to the constructor.  

I've updated this PR's description accordingly.

paulromano commented on include/enrico/surrogate_heat_driver.h:
```diff
@@ -3,18 +3,17 @@
-#include <mpi.h>
 
 #include <cstddef>
+#include <gsl/gsl>
```
This is not in the C++ standard library -- it belongs in the third-party library section.

RonRahaman replied:
Another question.  The core guidelines seem to always list headers with `< >` instead of `" "`, even for headers that aren't part of the standard library.  Should I follow that convention?

paulromano replied:
`< >` means "search for this file in the standard locations" whereas `" "` means "search for this file in the standard locations _and_ the present directory". The only case where `" "` is actually needed is if the file being included is truly in a relative location (that is not on the include path). For example, in "heat_fluids_driver.h", we have:
```C++
#include "driver.h"
```
If this were
```C++
#include <enrico/driver.h>
```
it would work because `include/enrico` is on the include path. With `" "` however we have to rely on it being relative.

paulromano replied:
From a practical standpoint, I usually use `< >` to indicate "external to this project" and `" "` to indicate "internal to this project".

paulromano commented on include/enrico/surrogate_heat_driver.h:
```diff
@@ -71,6 +74,14 @@ class SurrogateHeatDriver : public Driver {
   //!< temperature in [K] for each (axial segment, ring)
   xt::xtensor<double, 3> temperature_;
 
+  //!< density in [g/cm^3] for each (axial segment, ring)
```
(pin, axial segment, ring)

paulromano commented on include/enrico/surrogate_heat_driver.h:
```diff
@@ -71,6 +74,14 @@ class SurrogateHeatDriver : public Driver {
+  //!< density in [g/cm^3] for each (axial segment, ring)
+  xt::xtensor<double, 3> density_;
+
+  //! Value is 1 if (axial segment, ring) region is in fluid; otherwise 0
```
(pin, axial segment, ring)

paulromano commented on src/nek_driver.cpp:
```diff
@@ -1,24 +1,27 @@
-
 #include <climits>
 #include <fstream>
+#include <gsl/gsl>
```
Same thing here -- not standard library

paulromano commented:
Regarding pressure -- that will eventually be needed by a solver other than Nek5000. The subchannel solver that @sphamil put together needs the pressure at the channel exit (see Eq. 56 in our AD-SE-08-61 milestone report). Do we really want to leave it as a Nek5000-only input/field?

RonRahaman commented:
@paulromano If possible, I would also like to have a data member for pressure in the base class.  However, @aprilnovak pointed out that some solvers might need the pressure to be an array; while others might need the pressure to be a scalar.  In that case, the type of the data member in the base class would need to accommodate both those possibilities.  That was the motivation for moving pressure into the derived class.

If pressure were in the base class, I can think of a couple of ways to declare it.  Do either of these seem satisfactory to you?  
* Declare pressure as `xt::xtensor<double, 1>`.  For solvers that need a scalar pressure, treat it as a 1D array of shape (1,)
* Declare pressure as `xt::xarray<double>`.  For solvers that need a scalar pressure, it would be defined as a 0D array (which is how xtensor treats scalars, anyway).  For solvers that need an array, it would be defined as a 1D array.  

If either of these are satisfactory to you, then I would vote for putting pressure in the base class.

aprilnovak commented:
Here's how I interpret the temperature members of our classes - 

`CoupledDriver::temperature_` [public]: temperature of the most recent Picard iteration 
`HeatFluidsDriver::temperature()` [public]: _gets_ temperature from T/H solver
`NekDriver`: _not exposed directly_ - you have to get it by calling `NekDriver::temperature()`
`SurrogateHeatDriver`: [private (i.e. not exposed directly)] - you have to get it by calling `SurrogateHeatDriver::temperature()`

The temperatures known to the single-physics codes are never accessed directly, so there is no need for an `xt::xtensor<double, 1> temperature_` member in `HeatFluidsDriver`. 

No single-physics T/H data is actually owned by `HeatFluidsDriver`; rather, it's only grabbed by getter methods. So if we keep that design, I don't think it would make sense to have `HeatFluidsDriver` _own_ `xt::xtensor<double, 1> pressure_` (or whatever other type) because that would mean that instead of using getters to grab data from a T/H solver, we've put data in the base class. Hiding data in daughter classes of `HeatFluidsDriver` gives us flexibility of defining pressure as `double pressure_` or `xt::xtensor<double, 1> pressure_` or any other type needed internally. It also protects us from modifying the solutions from "black-box" T/H solvers by only letting us grab it.

We only really need a user-defined pressure to be provided to `NekDriver` as a means-to-an-end to calculate density. Even if the subchannel solver is incompressible, it will still involve a pressure solution (I'm assuming it won't be in nondimensional form), so it could have internal methods to return density based on IAPWS. The subchannel solver will have a pressure array used in its own internal calculations, but there's no need to expose it to the coupling classes because we really only need density and temperature for feedback.

Therefore, I think it makes more sense for `double pressure_` to only belong to `NekDriver` and to only be used in within the `xt::xtensor<double, 1> NekDriver::density()` method defined in `HeatFluidsDriver`. This would be consistent with hiding data in code-specific implementations and having the base class only define getter methods.

What do you think? If you think otherwise, then we should re-think how we currently only have getter methods in the `HeatFluidsDriver`.

paulromano commented:
Yes, I agree that the pressure solution for the subchannel solver is an implementation detail that is not important for other classes to know about and so we shouldn't be exposing it. The pressure I'm talking about though is really a pressure **boundary condition**. Even though the subchannel solver determines a pressure field, it has to know the pressure at the outlet to begin with. I think of the pressure being used by Nek5000 in the same way. Technically, when we're calculating density based on the Nek solution, we should account for the pressure solution, but in practice it won't appreciably change the densities being returned, so we decided to just use the pressure at the outlet, which is given as an input. So, my point is really that the pressure boundary condition is independent of the particular T-H solver. @aprilnovak given this consideration, do you have thoughts on where it should belong? Also, maybe we ought to change the input to `<pressure_bc>` and the field name to `pressure_bc_` or something similar.

aprilnovak commented:
> The pressure I'm talking about though is really a pressure boundary condition

Ohh, sorry for the confusion! Then yes, I think `pressure_bc_` should be added to the `HeatFluidsDriver`. I think we can set it as a `double` since the subchannel solver will likely set a uniform pressure outlet BC?

RonRahaman commented:
I am fine with having a `pressure_bc` member in the `HeatFluidsDriver` base class; and a `<pressure_bc>` element under the `<enrico>` node in `enrico.xml`.

aprilnovak commented:
The return type of `get_heat_driver` should now be `HeatFluidsDriver&` rather than `Driver&`

paulromano commented:
I'm satisfied with this PR -- @aprilnovak do you have any further comments?

aprilnovak commented:
Nope, looks good to me!
